### ISSUE TITLE
External hinting system for automatic layering

### ISSUE THREAD

**Issue by mikepurvis** [nlewo/nix2container#113]
Some limitations of the "popularity" based approach for automatic layering:

- It can only think in single store paths, rather than recognizing clusters that make sense to group together.
- It doesn't have any temporal context, for example optimizing blobs for how much their contents change over time.
- It doesn't have any awareness of dependency chains (other than for the popularity number).
- It doesn't account for the size of the store paths.

None of these are huge issues with "small" images, but they start to really limit the effectiveness of the layering once there are thousands of store paths going into an image.

One possible way to improve this situation would be to have some kind external scanner tool that could examine a bunch of related images, and maybe also instances of the images/closures over time, and produce an output that could be checked into source control and used to better optimize automatic layer generation for successive builds. By checking it in, builds remain pure and the developer is in control of how frequently to update the hint file (likely in conjunction with dependency changes or flake updates).

If there's interest in such a thing, perhaps this ticket can be a place to discuss what such a file could look like and how would be most effective to collect the data.

**Comment by mikepurvis** [comment]
🤖 Authored by Claude on Mike's behalf

Closing this out, ~two years on, having finally measured the thing I was speculating about. Full harness, corpora and results: **[nix-layer-corpus gist](https://gist.github.com/mikepurvis/f266e2128cb966301ca8d90a5602907c)**.

The premise has largely been overtaken. [NixOS/nixpkgs#355189](https://github.com/NixOS/nixpkgs/pull/355189) replaced popularity with a size-first algorithm and addressed three of the four weaknesses in my original post, and @aldoborrero has since ported it to Go here in [#204](https://github.com/nlewo/nix2container/pull/204), superseded by [#220](https://github.com/nlewo/nix2container/pull/220). So I measured the hint file against *that* baseline rather than against popularity, over 583 image-revisions spanning 52 weekly nixpkgs revisions (NixOS system closures, plus a set of Python apps), scored as simulated registry pull bytes across the year.

Three results:

1. **The size-first change is a big win.** Popularity comes in at 94% of simply shipping one single layer; the new algorithm halves that. #204 measured the same thing far more convincingly than I did — popularity's catch-all layer at ~71% of a 4 GB image, where touching one file re-uploaded ~2.9 GB against a real registry, dropping to ~9 MB after the change. Caveat on my number: I reimplemented popularity for the harness rather than calling nix2container's Go code, so read it as indicative of the scheme, not a measurement of this project.
2. **Co-occurrence hinting doesn't work** — neutral at generous layer caps, clearly worse at tight ones, in both the "subdivide by signature" and "merge matching roots" formulations. This is exactly the data `auto-layer.py`'s comments wish for, and supplying it didn't help.
3. **Temporal hinting is worth 0–4%**, and only at generous caps.

The reason is measurable: dependency *footprints* barely move (0.9–1.8% membership churn per week) while versions turn over constantly (33–39%), because a glibc or Python bump rebuilds everything downstream. A hint file would be encoding a near-constant.

One important caveat: **no corpus built from nixpkgs can really test this.** A nixpkgs package tracks an upstream release, so its dep set moves at upstream's pace. The case I actually filed this about — an actively developed workspace where a developer adds and drops deps between builds — doesn't exist anywhere in nixpkgs history. So the co-occurrence result is "not demonstrated on data that can show it" rather than refuted; testing it properly needs a real team's CI history.

What did turn up is on the algorithm side. Reading the actual groupings, the clustering is good but the layer *budget* is misallocated: the absorption rule degenerates at the root of an application image, so one layer routinely holds 21–57% of the closure, while trivial paths like `zlib` and `libffi` each hold a whole layer slot. Reclaiming slots from the smallest layers to subdivide the largest gives another **6–12%, using fewer layers** — and splitting by size does as well as splitting by churn, so it needs no external data at all. There's a consistent progression here: popularity's catch-all at ~71%, size-first's largest layer at 21–57%, and another 6–12% off that from spending the layer budget better. It should apply equally to the Go port in #204. Mine is a post-process, so the cleaner fix is probably inside the absorption rule itself — cap absorption at a fraction of closure size, or exclude the image root from being a primary. I haven't benchmarked it through `dockerTools`, only my own simulator.

#220 is the right shape for whatever remains of this. Making the layer split an input (`layersFile`) is exactly the artifact my results point at — not co-occurrence statistics, but a precomputed assignment. It also unblocks the one experiment I couldn't run: layer sharing today is accidental convergence, in that two images share a layer only when each independently derives an identical path set, and nothing coordinates boundaries across a set of images. Optimising that jointly needs the image set up front, which is something a vendor has and a zero-config default cannot — and with `layersFile` it can happen outside the tool entirely, with no algorithm to agree on in-tree.

So: I'd like to **close this**. The concrete complaints were addressed by 355189 and its port in #204, the hint file doesn't pay on any data constructible from nixpkgs, the remaining headroom I can find is an internal algorithmic fix needing no hinting infrastructure, and #220 covers the extension point far more cleanly than a bespoke hint format would.
